core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -161,51 +164,99 @@ class ZooKeeperClient(connectString: String,
+              callback(SetDataResponse(Code.get(rc), path, Option(ctx), stat, responseMetadata(sendTimeMs)))
+          }, ctx.orNull)
+        else
+          zooKeeper.multi(Seq(checkInfo.get.checkOp, Op.setData(path, data, version)).asJava, new MultiCallback {
+            override def processResult(rc: Int, multiOpPath: String, ctx: scala.Any, opResults: util.List[OpResult]): Unit = {
+              val (zkCheckResultCode, requestOpResult) = getMultiOpResults(opResults)
+              val stat = requestOpResult match {
+                case s: SetDataResult => s.getStat
+                case _ => null
+              }
+              callback(SetDataResponse(Code.get(rc), path, Option(ctx), stat, responseMetadata(sendTimeMs), zkCheckResultCode))
+            }
+          }, ctx.orNull)
+      case DeleteRequest(path, version, ctx, checkInfo) =>
+        if (checkInfo.isEmpty)
+          zooKeeper.delete(path, version, new VoidCallback {
+            override def processResult(rc: Int, path: String, ctx: Any): Unit =
+              callback(DeleteResponse(Code.get(rc), path, Option(ctx), responseMetadata(sendTimeMs)))
+          }, ctx.orNull)
+        else
+          zooKeeper.multi(Seq(checkInfo.get.checkOp, Op.delete(path, version)).asJava, new MultiCallback {
+            override def processResult(rc: Int, multiOpPath: String, ctx: scala.Any, opResults: util.List[OpResult]): Unit = {
+              val (zkCheckResultCode, _) = getMultiOpResults(opResults)
+              callback(DeleteResponse(Code.get(rc), path, Option(ctx), responseMetadata(sendTimeMs), zkCheckResultCode))
+            }
+          }, ctx.orNull)
+      case GetAclRequest(path, ctx, _) =>
         zooKeeper.getACL(path, null, new ACLCallback {
           override def processResult(rc: Int, path: String, ctx: Any, acl: java.util.List[ACL], stat: Stat): Unit = {
             callback(GetAclResponse(Code.get(rc), path, Option(ctx), Option(acl).map(_.asScala).getOrElse(Seq.empty),
               stat, responseMetadata(sendTimeMs)))
-        }}, ctx.orNull)
-      case SetAclRequest(path, acl, version, ctx) =>
+          }
+        }, ctx.orNull)
+      case SetAclRequest(path, acl, version, ctx, _) =>
         zooKeeper.setACL(path, acl.asJava, version, new StatCallback {
           override def processResult(rc: Int, path: String, ctx: Any, stat: Stat): Unit =
             callback(SetAclResponse(Code.get(rc), path, Option(ctx), stat, responseMetadata(sendTimeMs)))
         }, ctx.orNull)
     }
   }
 
+  private def getMultiOpResults(opResults: util.List[OpResult]): (Option[Code], OpResult) = {
+    require(opResults.size() == 2)
+    val checkOpResult = opResults.get(0)
+    val requestOpResult = opResults.get(1)
+
+    val checkResultCode = checkOpResult match {
+      case _: CheckResult => Some(Code.OK)
+      case e: ErrorResult => Some(Code.get(e.getErr))
+      case _ => None
```

Comment:
It seems that checkOpResult should be either `CheckResult` or `ErrorResult`. Maybe we should throw IllegalStateException otherwise? By doing so we could simplify the signature of `getMultiOpResults` to `(Code, OpResult)`. And we can also simplify the signature of e.g. `CreateResponse` such that zkVersionCheckResultCode is of type `Code`.

Reply:
Done.